src/Npgsql/NpgsqlReadBuffer.cs

```diff
@@ -30,33 +30,28 @@ public sealed partial class NpgsqlReadBuffer : IDisposable
 
         readonly Socket? _underlyingSocket;
 
-        CancellationTokenSource _timeoutCts = new CancellationTokenSource();
+        TimeoutCancellationTokenSourceWrapper _timeoutCts;
 
         /// <summary>
         /// Timeout for sync and async reads
         /// </summary>
         internal TimeSpan Timeout
         {
-            get => _currentTimeout;
+            get => _timeoutCts.Timeout;
             set
             {
-                if (_currentTimeout != value)
+                if (_timeoutCts.Timeout != value)
                 {
                     Debug.Assert(_underlyingSocket != null);
 
-                    _underlyingSocket.ReceiveTimeout = value > TimeSpan.Zero ? (int)value.TotalMilliseconds : -1;
-                    _currentTimeout = value;
+                    _underlyingSocket.ReceiveTimeout = value == System.Threading.Timeout.InfiniteTimeSpan
+                        ? 0
+                        : (int)value.TotalMilliseconds;
+                    _timeoutCts.Timeout = value;
                 }
             }
         }
 
-        /// <summary>
-        /// Contains the current value of the Socket's RecieveTimeout, used to determine whether
-        /// we need to change it when commands are received.
-        /// Also, used as a timeout for async reads.
-        /// </summary>
-        TimeSpan _currentTimeout;
```

Comment:
This is definitely something I do not like - we're relying on `TimeoutCancellationTokenSourceWrapper` to never change its `Timeout` property by itself. But I'm open to discussion.

Reply:
Well, in case of the buffers, since we want symmetry between the async and the sync API regarding timeouts, I don't see a reason why the instance of the stored `TimeoutCancellationTokenSourceWrapper` should have a different timeout than the instance of the stored `Socket`.

Reply:
They definitely shouldn't have a different timeout between each other. 
My only problem here is that the `Buffer` doesn't own the `TimeoutCancellationTokenSourceWrapper.Timeout` - only the `TimeoutCancellationTokenSourceWrapper` itself. The same can be said about the `Socket`.

But as this is more of an aesthetic problem (and no one else seems to mind), you can leave it as is.

Reply:
I see what you mean but in the actual implementation the `TimeoutCancellationTokenSourceWrapper` instance (`_timeoutCts`) is private and we only expose it's `Timeout` property via `NpgsqlReadBuffer.Timeout`, so it's really pure aesthetics.
The only way to make a mess here, is if you're coding inside the `NpgsqlReadBuffer` class.